File: base/src/main/java/com/github/ykiselev/base/game/client/Camera.java
package com.github.ykiselev.base.game.client;

import com.github.ykiselev.opengl.matrices.Matrix;
import com.github.ykiselev.opengl.matrices.Vector3f;
import com.github.ykiselev.opengl.pools.Vector3fPool;
import org.lwjgl.system.MemoryStack;

import java.nio.FloatBuffer;

public final class Camera {

    private double yaw, pitch;

    private float x, y, z, dx, dy, rx, ry, zNear = 0.1f, zFar = 100f, fow = 90f;

    private final Vector3f direction = new Vector3f(), up = new Vector3f(), right = new Vector3f();

    public void set(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public void move(float delta) {
        buildVectors();
        x += dx * delta;
        y += dy * delta;
    }

    public void strafe(float delta) {
        buildVectors();
        x += rx * delta;
        y += ry * delta;
    }

    public void moveUp(float delta) {
        z += delta;
    }

    private static double limitAngle(double value) {
        if (value > 360) {
            return value - 360;
        }
        if (value < -360) {
            return value + 360;
        }
        return value;
    }

    public void rotate(double dx, double dy) {
        yaw = limitAngle(yaw - dy);
        pitch = limitAngle(pitch + dx);
    }

    private void buildVectors() {
        try (var ms = MemoryStack.stackPush();
             var vectors = Vector3fPool.push()) {
            final FloatBuffer mat = ms.mallocFloat(16);

            buildRotation(mat);
            Matrix.inverse(mat, mat);

            direction.set(0, 0, -1);
            Matrix.multiply(mat, direction);
            up.set(0, 1, 0);
            Matrix.multiply(mat, up);
            right.set(1, 0, 0);
            Matrix.multiply(mat, right);

            Vector3f v = vectors.allocate();
            v.set(direction.x, direction.y, 0);
            v.normalize();
            dx = v.x;
            dy = v.y;

            v.set(right.x, right.y, 0);
            v.normalize();
            rx = v.x;
            ry = v.y;
        }
        direction.normalize();
        up.normalize();
        right.normalize();
    }

    private void buildRotation(FloatBuffer m) {
        Matrix.rotation(Math.toRadians(yaw - 90), 0, Math.toRadians(pitch), m);
    }

    public void apply(float ratio, FloatBuffer m) {
        Matrix.perspective((float) Math.toRadians(fow), ratio, zNear, zFar, m);

        try (MemoryStack ms = MemoryStack.stackPush()) {
            final FloatBuffer mat = ms.mallocFloat(16);

            buildRotation(mat);
            Matrix.multiply(m, mat, m);

            Matrix.identity(mat);
            Matrix.translate(mat, -x, -y, -z, mat);
            Matrix.multiply(m, mat, m);
        }
    }

    @Override
    public String toString() {
        return "Camera{" +
                "yaw=" + yaw +
                ", pitch=" + pitch +
                ", x=" + x +
                ", y=" + y +
                ", z=" + z +
                ", dx=" + dx +
                ", dy=" + dy +
                ", rx=" + rx +
                ", ry=" + ry +
                '}';
    }
}

File: base/src/main/java/com/github/ykiselev/base/game/client/Camera.kt
package com.github.ykiselev.base.game.client

import com.github.ykiselev.opengl.matrices.*

class Camera {

    private val direction = Vector3f()
    private val up = Vector3f()
    private val right = Vector3f()
    private var yaw = 0.0
    private var pitch = 0.0
    private var x = 0f
    private var y = 0f
    private var z = 0f
    private var dx = 0f
    private var dy = 0f
    private var rx = 0f
    private var ry = 0f
    private val zNear = 0.1f
    private val zFar = 100f
    private val fow = 90f

    fun set(x: Float, y: Float, z: Float) {
        this.x = x
        this.y = y
        this.z = z
    }

    fun move(delta: Float) {
        buildVectors()
        x += dx * delta
        y += dy * delta
    }

    fun strafe(delta: Float) {
        buildVectors()
        x += rx * delta
        y += ry * delta
    }

    fun moveUp(delta: Float) {
        z += delta
    }

    fun rotate(dx: Double, dy: Double) {
        yaw = limitAngle(yaw - dy)
        pitch = limitAngle(pitch + dx)
    }

    private fun buildVectors() {
        math {
            val mat = buildRotation().inverse()

            val newDirection = mat * vec3f(0f, 0f, -1f)
            //direction.set(0f, 0f, -1f)
            //multiply(mat, direction, direction)
            val newUp = mat * vec3f(0f, 1f, 0f)
            //up.set(0f, 1f, 0f)
            //multiply(mat, up, up)
            val newRight = mat * vec3f(1f, 0f, 0f)
            //right.set(1f, 0f, 0f)
            //multiply(mat, right, right)

            val v1 = vec3f(newDirection.x, newDirection.y, 0f).normalize()
            //val v1 = Vector3f(direction.x, direction.y, 0f).normalize()
            //v.normalize()
            dx = v1.x
            dy = v1.y

            val v2 = vec3f(newRight.x, newRight.y, 0f).normalize()
            //v.normalize()
            rx = v2.x
            ry = v2.y

            direction.set(newDirection.normalize())
            up.set(newUp.normalize())
            right.set(newRight.normalize())
        }
    }

    private fun MathArena.buildRotation(): Matrix =
        rotation(Math.toRadians(yaw - 90), 0.0, Math.toRadians(pitch))

    fun apply(ratio: Float, m: Matrix) {
        math {
            val p = perspective(Math.toRadians(fow.toDouble()).toFloat(), ratio, zNear, zFar)//, m!!)
            val pr = p * buildRotation()
            m.set(pr * identity().translate(-x, -y, -z))
        }
    }

    override fun toString(): String {
        return "Camera{" +
                "yaw=" + yaw +
                ", pitch=" + pitch +
                ", x=" + x +
                ", y=" + y +
                ", z=" + z +
                ", dx=" + dx +
                ", dy=" + dy +
                ", rx=" + rx +
                ", ry=" + ry +
                '}'
    }

    companion object {
        private fun limitAngle(value: Double): Double {
            if (value > 360) {
                return value - 360
            }
            if (value < -360) {
                return value + 360
            }
            return value
        }
    }
}
